File: ios/NetworkErrorMapper.swift
import Foundation
import ExpoModulesCore
import iOSNetworkContract

final class NetworkException: Exception {
    private let networkCode: String
    private let retryable: Bool
    private let httpStatus: Int?
    private let networkMessage: String?
    private let info: [String: Any]?

    init(
        code: String,
        retryable: Bool,
        httpStatus: Int? = nil,
        message: String? = nil,
        info: [String: Any]? = nil
    ) {
        self.networkCode = code
        self.retryable = retryable
        self.httpStatus = httpStatus
        self.networkMessage = message
        self.info = info
        super.init()
    }

    // The code field carries the full NetworkErrorPayload serialized as JSON.
    // The JS side parses e.code to reconstruct { code, retryable, httpStatus?, message?, info? }.
    override var code: String {
        var dict: [String: Any] = ["code": networkCode, "retryable": retryable]
        if let s = httpStatus { dict["httpStatus"] = s }
        if let m = networkMessage { dict["message"] = m }
        if let i = info, JSONSerialization.isValidJSONObject(i) { dict["info"] = i }
        guard
            let data = try? JSONSerialization.data(withJSONObject: dict),
            let str = String(data: data, encoding: .utf8)
        else { return networkCode }
        return str
    }

    override var reason: String { networkMessage ?? "Network error: \(networkCode)" }
}

enum NetworkErrorMapper {
    static func map(_ error: Error) -> NetworkException {
        if let networkError = error as? NetworkException { return networkError }

        // Typed error coming from the host's provider — pass it through verbatim.
        if let typed = error as? NetworkError {
            return NetworkException(
                code: typed.code,
                retryable: typed.retryable,
                httpStatus: typed.httpStatus,
                message: typed.message,
                info: typed.info
            )
        }

        // Structured-concurrency cancellation flows here when the request is aborted.
        if error is CancellationError {
            return NetworkException(code: "CANCELLED", retryable: false)
        }

        if let urlError = error as? URLError {
            if urlError.code == .cancelled {
                return NetworkException(code: "CANCELLED", retryable: false)
            }
            switch urlError.code {
            case .timedOut:
                return NetworkException(code: "TIMEOUT", retryable: true)
            case .notConnectedToInternet, .networkConnectionLost, .dataNotAllowed:
                return NetworkException(code: "NO_CONNECTIVITY", retryable: true)
            case .serverCertificateUntrusted,
                 .clientCertificateRejected,
                 .serverCertificateHasUnknownRoot,
                 .serverCertificateNotYetValid,
                 .serverCertificateHasBadDate:
                return NetworkException(code: "SSL_PINNING_FAILED", retryable: false)
            default:
                break
            }
        }

        let nsError = error as NSError
        if nsError.domain == "com.scotia.rnnetwork.http" {
            let status = nsError.code
            if (400...499).contains(status) {
                return NetworkException(code: "HTTP_CLIENT_ERROR", retryable: false, httpStatus: status)
            } else if (500...599).contains(status) {
                return NetworkException(code: "HTTP_SERVER_ERROR", retryable: true, httpStatus: status)
            }
        }

        return NetworkException(code: "UNKNOWN", retryable: false)
    }
}

File: android/src/main/java/expo/modules/rnnetwork/NetworkErrorMapper.kt
package expo.modules.rnnetwork

import com.scotia.rnnetwork.contracts.NetworkError
import expo.modules.kotlin.exception.CodedException
import kotlinx.coroutines.CancellationException
import org.json.JSONObject
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import javax.net.ssl.SSLException

// Builds the JSON payload the JS side reconstructs from `e.code`.
private fun buildCode(
    code: String,
    retryable: Boolean,
    httpStatus: Int?,
    message: String?,
    info: Map<String, Any?>?,
): String =
    try {
        val obj = JSONObject()
        obj.put("code", code)
        obj.put("retryable", retryable)
        httpStatus?.let { obj.put("httpStatus", it) }
        message?.let { obj.put("message", it) }
        info?.let { obj.put("info", JSONObject(it)) }
        obj.toString()
    } catch (e: Exception) {
        code
    }

class NetworkException(
    networkCode: String,
    retryable: Boolean,
    httpStatus: Int? = null,
    networkMessage: String? = null,
    info: Map<String, Any?>? = null,
) : CodedException(
    buildCode(networkCode, retryable, httpStatus, networkMessage, info),
    networkMessage ?: "Network error: $networkCode",
    null,
)

object NetworkErrorMapper {
    fun map(error: Throwable): NetworkException {
        if (error is NetworkException) return error

        // Typed error coming from the host's provider — pass it through verbatim.
        if (error is NetworkError) {
            return NetworkException(
                networkCode = error.code,
                retryable = error.retryable,
                httpStatus = error.httpStatus,
                networkMessage = error.message,
                info = error.info,
            )
        }

        return when (error) {
            // Coroutine cooperative cancellation flows here when the request is aborted.
            is CancellationException ->
                NetworkException("CANCELLED", retryable = false)
            is SSLException ->
                NetworkException("SSL_PINNING_FAILED", retryable = false)
            is SocketTimeoutException ->
                NetworkException("TIMEOUT", retryable = true)
            is UnknownHostException ->
                NetworkException("NO_CONNECTIVITY", retryable = true)
            is java.io.IOException -> {
                val message = error.message.orEmpty()
                val httpDomainStatus = message
                    .substringAfter("com.scotia.rnnetwork.http:", "")
                    .toIntOrNull()
                when {
                    httpDomainStatus != null -> when (httpDomainStatus) {
                        in 400..499 -> NetworkException("HTTP_CLIENT_ERROR", retryable = false, httpStatus = httpDomainStatus)
                        in 500..599 -> NetworkException("HTTP_SERVER_ERROR", retryable = true, httpStatus = httpDomainStatus)
                        else -> NetworkException("HTTP_ERROR", retryable = false, httpStatus = httpDomainStatus)
                    }
                    message.contains("timeout", ignoreCase = true) ->
                        NetworkException("TIMEOUT", retryable = true)
                    message.contains("unable to resolve", ignoreCase = true) ||
                    message.contains("no route to host", ignoreCase = true) ->
                        NetworkException("NO_CONNECTIVITY", retryable = true)
                    else ->
                        NetworkException("UNKNOWN", retryable = false)
                }
            }
            else -> NetworkException("UNKNOWN", retryable = false)
        }
    }
}
